### PR TITLE
Lido app/marcus/json

### PR DESCRIPTION
**Pull request by marhong2**
@tyleradams @jmorganmartin 

made it so we can switch if we use the JSON test or the default test. Please LMK if there is a better way for developers to switch tests. Otherwise, I have added in a function to save all of the tests.

### CODE REVIEW

**Comment by tyleradams** [comment]
Switch to JSON only.

File: test/formulas/date/date.js [hunk 1]
```diff
@@ -35,5 +35,5 @@ const parser = new FormulaParser({
 });
 
 describe('Date and Time Functions', function () {
```
Comment (tyleradams) [inline test/formulas/date/date.js:37]: Delete all these section specific ones, and just call generateTests once in one place.

File: test/utils.js [hunk 2]
```diff
@@ -1,26 +1,84 @@
 const expect = require('chai').expect;
+const fs = require('fs')
+const allTests = require('../TESTS.json')
+//true = use tests in js files
+//false = use tests in JSON
+const original = false
```
Comment (tyleradams) [inline test/utils.js:6]: Use JSON only. Delete old tests.

File: test/utils.js [hunk 3]
```diff
@@ -1,26 +1,84 @@
...
+                            else
+                                expect(result, `${formula} should equal ${expected}\n`).to.equal(expected);
+                        }
+                    })
+                })
+            })      
+        }
+        
+    },
+    saveTests: (TestCase, type) => {
```
Comment (tyleradams) [inline test/utils.js:61]: This is why JSON tests are AWESOME. You can programatically generate test answers into the test file itself.

File: test/utils.js [hunk 4]
```diff
@@ -1,26 +1,84 @@
...
+            const tests = allTests[type]
+            const funs = Object.keys(tests)
+            funs.forEach(fun => {
+                it(fun, () => {
+                    const formulas = Object.keys(tests[fun])
+                    formulas.forEach(formula => {
+                        let expected = tests[fun][formula]
+                        let result = parser.parse(formula, {row:1, col:1});
+                        if (result.result) result = result.result;
+                        //variable in nature, so I can not save them in a JSON
```
Comment (tyleradams) [inline test/utils.js:40]: Fake out the system clock. 
https://sinonjs.org/releases/v1.17.7/fake-timers/

File: test/utils.js [hunk 5]
```diff
@@ -1,26 +1,84 @@
...
+                        if(formula === "HOUR(NOW())")
+                            expected = new Date().getHours();
+                        if(formula === "YEAR(NOW())")
+                            expected = new Date().getFullYear();
+                        
+                        if (typeof result === "number" && typeof expected === "number") {
+                            expect(result, `${formula} should equal ${expected}\n`).to.closeTo(expected, 0.00000001);
+                        } else {
+                            // For FormulaError
+                            if (result.error)
```
Comment (tyleradams) [inline test/utils.js:50]: No nested ifs!
https://codefaster.substack.com/p/faster-if-statements

File: test/utils.js [hunk 6]
```diff
@@ -1,26 +1,84 @@
...
+                        let result = parser.parse(formula, {row:1, col:1});
+                        if (result.result) result = result.result;
+                        //variable in nature, so I can not save them in a JSON
+                        if(formula === "HOUR(NOW())")
+                            expected = new Date().getHours();
+                        if(formula === "YEAR(NOW())")
+                            expected = new Date().getFullYear();
+                        
+                        if (typeof result === "number" && typeof expected === "number") {
+                            expect(result, `${formula} should equal ${expected}\n`).to.closeTo(expected, 0.00000001);
```
Comment (tyleradams) [inline test/utils.js:47]: Also print result

File: test/utils.js (hunk 5 above)
Comment (tyleradams) [inline test/utils.js:50]: just have the expected have an error field 
```
{ "error": "!NA blah blah blah"}
```

File: test/utils.js [hunk 7]
```diff
@@ -1,26 +1,84 @@
...
+                        } else {
+                            // For FormulaError
+                            if (result.error)
+                                expect(expected === result.error, `${formula} should equal ${expected.error}\n`).to.equal(true);
+                            else
+                                expect(result, `${formula} should equal ${expected}\n`).to.equal(expected);
```
Comment (tyleradams) [inline test/utils.js:53]: Also print result

File: test/utils.js [hunk 8]
```diff
@@ -1,26 +1,84 @@
...
+        
+    },
+    saveTests: (TestCase, type) => {
+        const json = {},
+              functions = {},
+              funs = Object.keys(TestCase);
+        funs.forEach(fun => {
+            const formulaValuePairs = {};
+            funs.forEach(formula => {
+                const expected = TestCase[fun][formula];
```
Comment (tyleradams) [inline test/utils.js:68]: inline expected.

File: test/utils.js [hunk 9]
```diff
@@ -1,26 +1,84 @@
...
+                formulaValuePairs[formula] = expected
+            })
+            functions[fun] = formulaValuePairs
+        })
+        json[type] = formulaValuePairs;
+
+        fs.appendFileSync(file, JSON.stringify(json), function (err) {
+            if (err) throw err;
+            console.log('Saved!');
```
Comment (tyleradams) [inline test/utils.js:77]: No need, use rule of silence
http://www.catb.org/~esr/writings/taoup/html/ch01s06.html#id2878450

File: test/utils.js [hunk 10]
```diff
@@ -1,26 +1,84 @@
...
+        }
+        
+    },
+    saveTests: (TestCase, type) => {
+        const json = {},
+              functions = {},
+              funs = Object.keys(TestCase);
+        funs.forEach(fun => {
+            const formulaValuePairs = {};
+            funs.forEach(formula => {
```
Comment (jmorganmartin) [inline test/utils.js:67]: This looks like it might be a typo?

```suggestion
            fun.forEach(formula => {
```

**Review by jmorganmartin — CHANGES_REQUESTED** [review]
Tyler left a lot of good suggestions. I noticed a potential typo (could be wrong about that, though).

File: test/utils.js (hunk 6 above)
Comment (marhong2) [reply test/utils.js:47]: They currently do, e.g:
 (AssertionError: SUM(A1:A1 A1:A1) should equal 0: expected 1 to be close to 0 +/- 1e-8)
would you like me to print out result a second time?

Comment (tyleradams) [reply test/utils.js:47]: ah k, ty

**Pull request closed without merging**